sentinel-core/public/sph_u.cc

```diff
@@ -5,30 +5,34 @@
 #include "sentinel-core/common/string_resource_wrapper.h"
 #include "sentinel-core/public/sph_u.h"
 #include "sentinel-core/slot/base/slot_chain.h"
+#include "sentinel-core/slot/global_slot_chain.h"
 
 namespace Sentinel {
 
-EntryResult SphU::Entry(const EntryContextPtr& context, const std::string& r,
-                        EntryType t, int count, int flag) {
+EntryResult SphU::Entry(const EntryContextSharedPtr& context,
+                        const std::string& r, EntryType t, int count,
+                        int flag) {
   auto resource = std::make_shared<StringResourceWrapper>(r, t);
-  EntrySharedPtr e = std::make_shared<Sentinel::Entry>(resource);
-  // Set current entry to new created context.
-  context->set_cur_entry(e);
+  EntrySharedPtr e = std::make_shared<Sentinel::Entry>(resource, context);
 
-  Slot::SlotChainSharedPtr chain;  // TODO: get the global chain
-  Stat::NodePtr empty_node = nullptr;
-  auto result = chain->Entry(context, resource, empty_node, count, flag);
+  Slot::SlotChainSharedPtr chain = Slot::GlobalSlotChain;
+  if (chain == nullptr) {
+    // TODO: should warn here.
+    return EntryResult{e};
+  }
+  Stat::NodeSharedPtr empty_node = nullptr;
+  auto result = chain->Entry(e, resource, empty_node, count, flag);
   if (result->status() == Slot::TokenStatus::RESULT_STATUS_BLOCKED) {
-    e->Exit(count);
+    EntryResult{e}.Exit();
```

Comment:
I prefer `e->Exit ()`, I don't know why build an `EntryResult` here

Reply:
The `Exit` method has been moved from `Entry` class to `EntryResult` method to avoid potential circular dependency problem, so we have to exit via the `EntryResult` class.

Reply:
Whether can replace with `chain->Exit(e, resource,....)`?

Reply:
Yes, till now it's OK, but we need to keep the behavior consistent in both two locations. If we modify the logic in `EntryResult::Exit`, we should also modify here. So we can replace the code with `chain->Exit(e, resource,....)` and add comments to inform this.